src/linker/Linker.Steps/MarkSubStepsDispatcher.cs

```diff
@@ -0,0 +1,192 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+
+using Mono.Cecil;
+using Mono.Collections.Generic;
+
+namespace Mono.Linker.Steps
+{
+	//
+	// Dispatcher for SubSteps which only need to run on marked assemblies.
+	// This simplifies the implementation of linker custom steps, in the same
+	// way that SubStepsDispatcher does, but it implements IMarkHandler
+	// and registers a callback that gets invoked during MarkStep when an
+	// assembly gets marked.
+	//
+	public abstract class MarkSubStepsDispatcher : IMarkHandler
+	{
+		readonly List<ISubStep> substeps;
+
+		List<ISubStep> on_assemblies;
+		List<ISubStep> on_types;
+		List<ISubStep> on_fields;
+		List<ISubStep> on_methods;
+		List<ISubStep> on_properties;
+		List<ISubStep> on_events;
+
+		protected MarkSubStepsDispatcher ()
+		{
+			substeps = new List<ISubStep> ();
+		}
+
+		protected MarkSubStepsDispatcher (IEnumerable<ISubStep> subSteps)
+		{
+			substeps = new List<ISubStep> (subSteps);
+		}
+
+		public void Add (ISubStep substep)
+		{
+			substeps.Add (substep);
+		}
```

Comment:
I'd remove this method and default ctor to avoid problems with calling Add at the right time.